### DISCUSSION TITLE
Field time optional

### DISCUSSION THREAD

**Discussion by mguyard** [pydantic/pydantic#9284] [Pydantic V2]
Hello,

I try to use time type as optional. Time is sometime present and sometime not as it's depend of type content.
```
class NightConfig(BaseModel):
    type: Literal["off", "time", "sunset", "lux", "prefer-lux"]
    time: Optional[time] = None

    @model_validator(mode='before')
    def checks(cls, values):
        if values['type'] in ["time","prefer-lux"] and values.get('time') is None:
            raise ValueError("time argument must be defined when type is time or prefer-lux")
        return values
```

This code don't work as expected.
It's the same with 
`time: Optional[time]`
or
`time: time | None = None`

For the first code, i've this error :

> 2024-04-19 13:43:50.295902 ERROR CoversManager: 1 validation error for Config
> common.night.time
>  Input should be None [type=none_required, input_value=82800, input_type=int]
>    For further information visit https://errors.pydantic.dev/2.7/v/none_required

For the second :
> 2024-04-19 13:53:11.804003 ERROR CoversManager: 1 validation error for Config
> common.night.time
>  Field required [type=missing, input_value={'type': 'off'}, input_type=dict]
>    For further information visit https://errors.pydantic.dev/2.7/v/missing

and the third :
>   File "/config/apps/CoversManager/src/config_validator.py", line 41, in MorningConfig
>     time: time | None = None
>          ~~~~~^~~~~~
> TypeError: unsupported operand type(s) for |: 'NoneType' and 'NoneType'

As a newbie of Pydantic, i'm a little bit lost on how to do this. If someone can help me this will be very appreciated.

**Comment by Viicos** [comment]
Your type annotation clashes with the field name (`time: time | None`). This is a known issue that was reported [here](https://github.com/pydantic/pydantic/issues/7327), but is specific to how Python resolves annotations. [This rule](https://github.com/Viicos/flake8-pydantic?tab=readme-ov-file#pyd005---field-name-overrides-annotation) from the `flake8-pydantic` plugin can help catch these issues.

**Comment by mguyard** [comment]
Hello @Viicos,

Thank you. This works perfectly.
